src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,186 @@ mod type_keyword {}
+///
+/// # Unsafe abilities
+///
+/// **No matter what, Safe Rust can't cause Undefined Behavior**. This is
+/// referred to as [soundness]: a well-typed program actually has the desired
+/// properties. The [Nomicon][nomicon-soundness] has a more detailed explanation
+/// on the subject.
+///
+/// To ensure soundness, Safe Rust is restricted enough that it can be
+/// automatically checked. Sometimes, however, it is necessary to write code
+/// that is correct for reasons which are too clever for the compiler to
+/// understand. In those cases, you need to use Unsafe Rust. Here are the
+/// abilities Unsafe Rust has in addition to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Implement `unsafe` [`trait`]s
+/// - Call `unsafe` functions
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// However, this extra power comes with extra responsibilities: it is now up to
+/// you to ensure soundness. The `unsafe` keyword helps by clearly marking the
+/// pieces of code that need to worry about this.
+///
+/// ## The different meanings of `unsafe`
+///
+/// Not all uses of `unsafe` are equivalent: some are here to mark the existence
+/// of a contract the programmer must check, others are to say "I have checked
+/// the contract, go ahead and do this". The following
+/// [discussion on Rust Internals] has more in-depth explanations about this but
+/// here is a summary of the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait`: implementing the [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the contract necessary to call the operations inside the
+/// block has been checked and is guaranteed to be respected.
+/// - `unsafe impl`: the contract necessary to implement the trait has been
+/// checked and is guaranteed to be respected.
+///
+/// As of now (Rust 1.44), `unsafe fn` also acts like an `unsafe {}` block
+/// around the code inside the function. This means it is not just a signal to
+/// the caller, but also promises that the preconditions for the operations
+/// inside the function are upheld. Mixing these two meanings can be confusing
+/// and [proposal]s exist to use `unsafe {}` blocks inside such functions when
+/// making `unsafe` operations.
+///
+/// See the [Rustnomicon] and the [Reference] for more informations.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// # Examples
+///
+/// ## Marking elements as `unsafe`
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
```

Comment:
It would be good to expand on this being very easily an anti-pattern: a `trait` with an `unsafe fn` associated item must be `unsafe` itself, unless the `unsafe fn` cannot be overriden (blanket `impl` for all possible implementors of the trait).

Otherwise the `unsafe fn` associated function becomes unusable within a non-`unsafe` generic function context, since each implementor is free to choose the narrow contract they want for that function.

  - Example

    ```rust
    trait DerefRead {
        unsafe fn deref_read (_: *const Self) -> Self
        ;
    }
    // Nothing prevents someone from doing
    impl DerefRead for ... {
        unsafe fn deref_read (_: *const Self) -> Read
        {
            ::core::ptr::null().read() // always UB
        }
    }  
    ```
---

Small nit: it's not a `method` but a `function`, here.

Reply:
I don't exactly understand why having an `unsafe fn` should make the trait `unsafe`. The original documentation on the trait should certainly provide the documentation about the safety requirements on the function/method, but I don't see how it is linked to the whole trait.

I don't think I understand your example: to me it just shows the trait has been implemented very incorrectly, nothing about the whole trait being unsafe.

Reply:
> it just shows the trait has been implemented very incorrectly, nothing about the whole trait being unsafe.

If implementing the trait incorrectly could cause UB, then it should be unsafe to implement. That's the _meaning_ of an unsafe trait.

Reply:
Oh yeah I wrote that completely wrong 🤦.

I meant to write the *function* is implemented incorrectly.

I found some examples of safe traits containing unsafe functions (though I did not review the links I list thoroughly so maybe its is a bad design decision for them):

- [Rustr](https://github.com/joelwkall/rustr/blob/20ad63cab0f36e69da8878175f91e7f18dd4e188/src/traits/mod.rs#L58)
- [RustAlgebloat](https://github.com/SiegeLord/RustAlgebloat/blob/c12715fb5b4d0e190cb48b4ae0c2d9af815c5591/algebloat/src/traits.rs#L18)
- [fuchsia](https://github.com/mehulagg/fuchsia/blob/3f56175ee594da6b287d5fb19f2f0eccea2897f0/third_party/rust_crates/vendor/ppv-lite86/src/types.rs#L60)

Reply:
Indeed that "mistake" is made out in the wild! And that's why I think it would be a good thing to document about this issue, to better raise awareness about it 🙂

Reply:
I am really reluctant on adding more and more bloat to the docs for the `unsafe` keyword. They are already very heavy and details like this belongs to the book/the nomicon I think, not the keyword doc.

I could be convinced of adding it, but I don't know where to add without making it feel long and focusing too much on small details. Maybe it's time for a `# Common mistakes` section ?

Reply:
FWIW I agree, I think a lot of the recent suggestions would be great to add to the nomicon but not here.